global_shortcuts.go

```diff
@@ -0,0 +1,135 @@
+	EventNameGlobalShortcutsEventUnregisteredAll = "global.shortcuts.event.unregistered.all"
+	EventNameGlobalShortcutEventTriggered        = "global.shortcuts.event.triggered"
+)
+
+type globalShortcutsCallback func()
+
+// GlobalShortcuts represents a global shortcut
+type GlobalShortcuts struct {
+	*object
+	m         *sync.Mutex
+	callbacks map[string]*globalShortcutsCallback // Store all registered Global Shortcuts
+}
+
+func newGlobalShortcuts(ctx context.Context, d *dispatcher, i *identifier, w *writer) (gs *GlobalShortcuts) {
+
+	gs = &GlobalShortcuts{object: newObject(ctx, d, i, w, i.new()), m: new(sync.Mutex), callbacks: make(map[string]*globalShortcutsCallback)}
+	gs.On(EventNameGlobalShortcutEventTriggered, func(e Event) (deleteListener bool) { // Register the listener for the triggered event
+		gs.execCallback(e.GlobalShortcuts.Accelerator)
+		return
+	})
+	return
+}
+
+// Register Register global shortcuts
+func (gs *GlobalShortcuts) Register(accelerator string, callback globalShortcutsCallback) (isRegistered bool, err error) {
+
+	if err = gs.ctx.Err(); err != nil {
+		return
+	}
+
+	// Send an event to astilectron to register the global shortcut
+	var event = Event{Name: EventNameGlobalShortcutsCmdRegister, TargetID: gs.id, GlobalShortcuts: &EventGlobalShortcuts{Accelerator: accelerator}}
+	result, err := synchronousEvent(gs.ctx, gs, gs.w, event, EventNameGlobalShortcutsEventRegistered)
+
+	if err != nil {
+		return
+	}
+
+	// If registered successfully, add the callback to the map
+	if result.GlobalShortcuts.IsRegistered {
+		gs.m.Lock()
+		gs.callbacks[accelerator] = &callback
+		gs.m.Unlock()
+	}
+
+	isRegistered = result.GlobalShortcuts.IsRegistered
+	return
+}
+
+// IsRegistered Check if a global shortcut is registered
+func (gs *GlobalShortcuts) IsRegistered(accelerator string) (isRegistered bool, err error) {
+
+	if err = gs.ctx.Err(); err != nil {
+		return
+	}
+
+	// Send an event to astilectron to check if global shortcut is registered
+	var event = Event{Name: EventNameGlobalShortcutsCmdIsRegistered, TargetID: gs.id, GlobalShortcuts: &EventGlobalShortcuts{Accelerator: accelerator}}
+	result, err := synchronousEvent(gs.ctx, gs, gs.w, event, EventNameGlobalShortcutsEventIsRegistered)
+
+	if err != nil {
+		return
+	}
+
+	isRegistered = result.GlobalShortcuts.IsRegistered
+	return
+}
+
+// Unregister Unregister a global shortcut
+func (gs *GlobalShortcuts) Unregister(accelerator string) (err error) {
+
+	if err = gs.ctx.Err(); err != nil {
+		return
+	}
+
+	// Send an event to astilectron to unregister the global shortcut
+	var event = Event{Name: EventNameGlobalShortcutsCmdUnregister, TargetID: gs.id, GlobalShortcuts: &EventGlobalShortcuts{Accelerator: accelerator}}
+	_, err = synchronousEvent(gs.ctx, gs, gs.w, event, EventNameGlobalShortcutsEventUnregistered)
+
+	if err != nil {
+		return
+	}
+
+	// No need to find the callback from the map and delete it
```

Comment:
I do think cleaning the map is still a good idea

Reply:
The reason why I didn't clean the map is that it is an extra cost to determine whether the shortcut should be removed from the map. For example, a user registered an event with `Ctrl+Shift+S` then unregistered with `S+Shift+Ctrl`, expected `Ctrl`+`Shift`+`S` should be unregistred. Therefore `Ctrl+Shift+S` should be removed from the map.
A more complex case: registered with `CtrlOrCmd+X` then unregistered with `Cmd+X`, expected `Ctrl`+`X` is unregistered. Therefore `CtrlOrCmd+X` should be removed from the map.
It's too complex to determine all combinations, the best way is to let electron decide whether the shortcuts should be registered or unregistered.

Reply:
Mmm last time I used shortcuts in JS, `Ctrl+Shift+S` was different than `S+Shift+Ctrl` 🤔  Therefore wanting to register a shortcut `Ctrl+Shift+S` and unregister it with `S+Shift+Ctrl` should not work in Electron 🤔 

Nevertheless, I think it's not too much to ask the developer to be consistent in the accelerator he/she uses when registering/unregistering a shortcut. Also I'd rather keep the map the closest to reality.

Therefore, I'd rather see the map cleaned 👍

Reply:
> Mmm last time I used shortcuts in JS, Ctrl+Shift+S was different than S+Shift+Ctrl 🤔 Therefore wanting to register a shortcut Ctrl+Shift+S and unregister it with S+Shift+Ctrl should not work in Electron 🤔

Howerver, the combination of `ctrl`+`shift`+`S` was unregistered successfully.

> Therefore, I'd rather see the map cleaned 👍

Sure thing